include/dlaf/common/assert.h

```diff
@@ -0,0 +1,38 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2019, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#pragma once
+
+#include <exception>
+#include <iostream>
+
+#include "dlaf/common/utils.h"
+
+#define DLAF_CHECK(category, condition, ...)                                                           \
+  if (!(condition)) {                                                                                  \
+    std::cerr << "[" category "-ERROR] " << __FILE__ << ":" << __LINE__ << " : " << DLAF_FUNCTION_NAME \
```

Comment:
Is `DLAF_FUNCTION_NAME` the name of the function calling the macro? Do we need that?

The string representation of the expression is missing from the string: `#condition`. I would put that instead of the function name.

Reply:
`DLAF_FUNCTION_NAME` is set from CMake, and it is:
- `__PRETTY_FUNCTION__` if available
- `__func__` otherwise

https://github.com/eth-cscs/DLA-Future/blob/aece6d0b770f6bbaff02f9e64bbb62c781ebdad2/src/CMakeLists.txt#L47-L55

So we can use the GCC extension `__PRETTY_FUNCTION__` when available. We can decide to use just the C++ standard `__func__`. Vaguely related to this, I have on my to-do list to anticipate [`source_location`](https://en.cppreference.com/w/cpp/utility/source_location) from C++20.

About the `#condition`, it is in the next line. See next examples.

## DEFAULT MESSAGE
```cpp
DLAF_ASSERT((false == true && index != common::Index2D<int, TAG>{}));
```
```bash
[ASSERT-ERROR] /Users/ialberto/workspace/repos/DLA-Future/example_assert.cpp:11 : int main()
(false == true && index != common::Index2D<int, TAG>{})
```

## CUSTOM MESSAGE
```cpp
DLAF_ASSERT((false == true && index != common::Index2D<int, TAG>{}), "Custom message", index);
```
```bash
[ASSERT-ERROR] /Users/ialberto/workspace/repos/DLA-Future/example_assert.cpp:12 : int main()
(false == true && index != common::Index2D<int, TAG>{}) Custom message (1, 2)
```